enhancements/network/vrf-cni.md

```diff
@@ -0,0 +1,123 @@
+
+### Goals
+
+Being able to assign a secondary interface to a particular vrf, and eventually having more secondary interfaces of the same pod assigned to the same VRF.
+
+### Non-Goals
+
+Assigning host level interfaces to VRFs.
+
+## Proposal
+
+The proposal is to introduce an additional CNI plugin that takes the vrf name as an input, and can be chained to the “main” cni used for the secondary interface, and that has the effect of creating the requested vrf (if it does not exist) and to add the interface to that vrf.
+
+In this way, the network interface returned by any CNI can be isolated in its own VRF, satisfying the above requirements without the need of additional capabilities.
+
+### User Stories [optional]
+
+#### Story 1
+
+As a user, I want to be able to optionally specify a VRF which the secondary nic is attached to.
+
+### Implementation Details/Notes/Constraints
+
+By adding several networks with the same `vrf` name, the CNI will ensure that they are all added to the same vrf, creating it only the first time.
+
+A `NetworkAttachmentDefinition` CNI config of `NetworkAttachmentDefinition` would look like
+
+```json
+{
+    "cniVersion": "0.3.1",
+    "name": "macvlan-vrf",
+    "plugins": [
+      {
+        "type": "macvlan",
+        "mode": "bridge",
+        "ipam": {
+          "type": "host-local",
+          "subnet": "192.168.1.0/24",
+          "rangeStart": "192.168.1.200",
+          "rangeEnd": "192.168.1.216",
+        }
+      },
+      {
+        "type": "vrf",
+        "vrfname": "blue"
+      }
+    ]
+}
+```
+
+This can be supported at `NetworkAttachmentDefinition` level, or by documenting (and validating) chained CNIs in the `Network` CRD of `operator.openshift.io/v1`.
+
+At the same time, the `SriovNetwork` type needs to be extended for chaining meta CNIs.
+
+### Risks and Mitigations
+
+## Design Details
+
+This would be shipped as a new CNI plugin, ideally being part of the container networking plugins [repo](https://github.com/containernetworking/plugins),
+which is then consumed in our [openshift variant](https://github.com/openshift/containernetworking-plugins). If not feasible, other alternatives
+may be considered for shipping it.
+
+The CNI implementation is pretty simple: it takes the result of the main CNI, creates the VRF in the pod's namespace and assignes the interface to the VRF.
+
+If the VRF already exists, it only adds the interface. When removing an interface, it does the same, checking if the interface is the last removed from the
+VRF and eventually delete the VRF.
+
+### Test Plan
+
+Apart from local unit tests, e2e tests will be added in the [cnf-features-deploy](https://github.com/openshift-kni/cnf-features-deploy) e2e integration test suite.
+Basic tests can be added under openshift/origin test suite too.
+
+### Graduation Criteria
```

Comment:
The `Graduation Criteria` defined both `dev preview` to `tech preview`  and `tech preview` to `GA`.
Which level of support we'd like to have for the initial vrf release? Is it Tech Preview? we might just need one.

Reply:
I'd say tech preview, but I thought it was good to have the next steps too.
So, the idea is to have only one and remove the other?

Reply:
I might be wrong, my understanding is we need the one that will be implemented in next release if it is necessary to go with several stages (DP -> TP -> GA).  If several stages are defined, then we probably want to highlight which stage this enhancement doc is targeting and have another enhancement (or update this enhancement) when it will become GA. 

Let's hear more from other reviewers as well.

Reply:
Ok, I'll leave as it is for now